internal/cortex/frontend/transport/handler.go

```diff
@@ -55,20 +58,51 @@ type Handler struct {
 	cfg          HandlerConfig
 	log          log.Logger
 	roundTripper http.RoundTripper
+	lru          *lru.Cache
+	regex        *regexp.Regexp
+	errorExtract *regexp.Regexp
 
 	// Metrics.
 	querySeconds *prometheus.CounterVec
 	querySeries  *prometheus.CounterVec
 	queryBytes   *prometheus.CounterVec
+	totalQueries prometheus.Counter
+	cachedHits   prometheus.Counter
 	activeUsers  *util.ActiveUsersCleanupService
 }
 
+// CacheableError Returns true if response code is in pre-defined cacheable errors list, else returns false
+func CacheableError(statusCode int) bool {
+	for _, errStatusCode := range cacheableResponseCodes {
+		if errStatusCode == statusCode {
+			return true
+		}
+	}
+	return false
+}
+
 // NewHandler creates a new frontend handler.
 func NewHandler(cfg HandlerConfig, roundTripper http.RoundTripper, log log.Logger, reg prometheus.Registerer) http.Handler {
+	var (
+		LRU *lru.Cache
+	)
+
+	if cfg.FailedQueryCacheCapacity > 0 {
+		LRU_res, err := lru.New(cfg.FailedQueryCacheCapacity)
+		LRU = LRU_res
+		if err != nil {
+			LRU = nil
+			level.Error(log).Log("msg", "failed to create lru cache", "err", err)
+		}
+	}
+
 	h := &Handler{
 		cfg:          cfg,
 		log:          log,
 		roundTripper: roundTripper,
+		lru:          LRU,
+		regex:        regexp.MustCompile(`[\s\n\t]+`),
+		errorExtract: regexp.MustCompile(`Code\((\d+)\)`),
 	}
 
 	if cfg.QueryStatsEnabled {
```

Comment:
Move the counter out of this `if`. `QueryStatsEnabled` is not enabled by default.